kernel/free_list.cpp

```diff
@@ -0,0 +1,206 @@
+bool is_the_first_in_buddy(const memory::page_frame_t *frame, size_t order);
+
+} // namespace
+
+
+bool memory::allocators::FreeList::initialize(size_t order,
+		bool disable_squashing) {
+	list_ = nullptr;
+	order_ = order;
+	disable_squashing_ = disable_squashing;
+	return true;
+}
+
+memory::page_frame_t *memory::allocators::FreeList::get() {
+	item_type *free_item = list_;
+	if (free_item) {
+		list_ = list_->next;
+		free_item->next = nullptr;
+		--stats.items;
+	}
+	sanity_check();
+	return reinterpret_cast<page_frame_t *>(free_item);
+}
+
+
+memory::page_frame_t *memory::allocators::FreeList::put(page_frame_t *frame) {
+	auto new_item = reinterpret_cast<item_type *>(frame);
+	new_item->next = nullptr;
+
+	if (list_ == nullptr) {
+		list_ = new_item;
+		++stats.items;
+		sanity_check();
+		return nullptr;
+	}
+
+	item_type *last_lesser_item;
+	item_type *prelast_lesser_item;
+	find_last_lesser(new_item, last_lesser_item, prelast_lesser_item);
+
+	if (disable_squashing_) {
+		if (last_lesser_item == nullptr) {
+			new_item->next = list_;
+			list_ = new_item;
+		} else {
+			new_item->next = last_lesser_item->next;
+			last_lesser_item->next = new_item;
+		}
+		++stats.items;
+		sanity_check();
+		return nullptr;
+	}
+
+	auto last_lesser_frame =
+			reinterpret_cast<page_frame_t *>(last_lesser_item);
+
+	if (last_lesser_item == nullptr) {
+		// Element should be placed to the beginning of the list.
+		if (are_consequent(frame,
+				reinterpret_cast<page_frame_t *>(list_),
+				order_)
+				&& is_the_first_in_buddy(frame, order_)) {
+			list_ = list_->next;
+			--stats.items;
+			sanity_check();
+			return frame;
+		} else {
+			new_item->next = list_;
+			list_ = new_item;
+			++stats.items;
+			sanity_check();
+			return nullptr;
+		}
+	}
+
+	auto next_item = last_lesser_item->next;
+	auto next_frame = reinterpret_cast<page_frame_t *>(next_item);
+
+	if (are_consequent(last_lesser_frame, frame, order_)
+			&& is_the_first_in_buddy(last_lesser_frame, order_)) {
+		if (prelast_lesser_item) {
+			prelast_lesser_item->next = last_lesser_item->next;
+			last_lesser_item->next = nullptr;
+			--stats.items;
+			sanity_check();
+			return last_lesser_frame;
+		} else {
+			list_ = last_lesser_item->next;
+			last_lesser_item->next = nullptr;
+			--stats.items;
+			sanity_check();
+			return last_lesser_frame;
+		}
+	}
+
+	if (are_consequent(frame, next_frame, order_)
+			&& is_the_first_in_buddy(frame, order_)) {
+		last_lesser_item->next = next_item->next;
+		--stats.items;
+		sanity_check();
+		return frame;
+	}
+
+	new_item->next = last_lesser_item->next;
+	last_lesser_item->next = new_item;
+	++stats.items;
+	sanity_check();
+	return nullptr;
+}
+
+
+void memory::allocators::FreeList::find_last_lesser(const item_type *new_item,
+		item_type * &last_lesser, item_type * &prelast_lesser) const {
+	prelast_lesser = nullptr;
+	last_lesser = nullptr;
+	for (auto item = list_; item && item < new_item; item = item->next) {
```

Comment:
:grey_exclamation: `new_item` is not nullptr; therefore:
if `item == nullptr` is truth; then `item < new_item` is also truth.
In reverse, if `item == nullptr` is false; then result of whole experssion is the same as result of `item < new_item`.
Thus checking of `item` for being `nullptr` can be deleted.